test/device_selection_tests.cc

```diff
@@ -0,0 +1,641 @@
+	celerity::detail::config cfg(nullptr, nullptr);
+
+	SECTION("pick_device prefers a particular device over all") {
+		mock_platform tp_0(0);
+		mock_device td_1(0, tp_0, cl::sycl::info::device_type::cpu);
+		tp_0.set_devices({td_1});
+		tp_0.set_info("foo_0");
+
+		mock_platform tp_1(1);
+		mock_device td_2(0, tp_1, cl::sycl::info::device_type::gpu);
+		mock_device td_3(1, tp_1, cl::sycl::info::device_type::gpu);
+		mock_device td_4(2, tp_1, cl::sycl::info::device_type::gpu);
+		mock_device td_5(3, tp_1, cl::sycl::info::device_type::gpu);
+		tp_1.set_devices({td_2, td_3, td_4, td_5});
+		tp_1.set_info("foo_1");
+
+		mock_platform tp_2(2);
+		mock_device td_6(0, tp_2, cl::sycl::info::device_type::accelerator);
+		mock_device td_7(1, tp_2, cl::sycl::info::device_type::accelerator);
+		mock_device td_8(2, tp_2, cl::sycl::info::device_type::accelerator);
+		mock_device td_9(3, tp_2, cl::sycl::info::device_type::accelerator);
+		tp_2.set_devices({td_6, td_7, td_8, td_9});
+		tp_2.set_info("foo_2");
+
+		auto device_selector = [td_7](const mock_device& d) -> int { return d == td_7 ? 2 : 1; };
+
+		auto device = pick_device(cfg, device_selector, std::vector<mock_platform>{tp_0, tp_1, tp_2});
+		std::vector<mock_device> devices = tp_2.get_devices();
+		CHECK(device == devices[1]);
+	}
+
+	SECTION("pick_device prefers a group of devices") {
+		log_capture lc;
+		mock_platform tp_0(0);
+		mock_device td_1(0, tp_0, cl::sycl::info::device_type::cpu);
+		tp_0.set_devices({td_1});
+		tp_0.set_info("foo_0");
+
+		mock_platform tp_1(1);
+		mock_device td_2(0, tp_1, cl::sycl::info::device_type::gpu);
+		mock_device td_3(1, tp_1, cl::sycl::info::device_type::gpu);
+		mock_device td_4(2, tp_1, cl::sycl::info::device_type::gpu);
+		tp_1.set_devices({td_2, td_3, td_4});
+		tp_1.set_info("foo_1");
+
+		mock_platform tp_2(2);
+		mock_device td_6(0, tp_2, cl::sycl::info::device_type::gpu);
+		mock_device td_7(1, tp_2, cl::sycl::info::device_type::gpu);
+		mock_device td_8(2, tp_2, cl::sycl::info::device_type::accelerator);
+		mock_device td_9(3, tp_2, cl::sycl::info::device_type::accelerator);
+		mock_device td_10(4, tp_2, cl::sycl::info::device_type::accelerator);
+		mock_device td_11(5, tp_2, cl::sycl::info::device_type::accelerator);
+		tp_2.set_devices({td_6, td_7, td_8, td_9, td_10, td_11});
+		tp_2.set_info("foo_2");
+
+		size_t node_count = 4;
+		size_t local_rank = 3;
+		size_t local_num_cpus = 1;
+
+		celerity::detail::host_config h_cfg{node_count, local_rank, local_num_cpus};
+		celerity::detail::config_testspy::set_mock_host_cfg(cfg, h_cfg);
+
+		auto device_selector = [](const mock_device& d) -> int { return d.get_type() == cl::sycl::info::device_type::accelerator ? 2 : 1; };
+
+		auto device = pick_device(cfg, device_selector, std::vector<mock_platform>{tp_0, tp_1, tp_2});
+		std::vector<mock_device> devices = tp_2.get_devices();
+		CHECK_THAT(lc.get_log(), Catch::Matchers::ContainsSubstring("device selector specified: platform"));
```

Comment:
Why are you not checking for the full string here?

Reply:
At this point, I thought since I am anyway checking for the right device the whole message isn't necessary, but you're right it does look uncanny, I have now changed it to the full message.